core/editor/src/tool/tools/fill.rs

```diff
@@ -0,0 +1,34 @@
+use crate::message_prelude::*;
+use crate::tool::ToolActionHandlerData;
+use document_core::Operation;
+
+#[derive(Default)]
+pub struct Fill;
+
+#[impl_message(Message, ToolMessage, Fill)]
+#[derive(PartialEq, Clone, Debug)]
+pub enum FillMessage {
+	MouseDown,
+}
+
+impl<'a> MessageHandler<ToolMessage, ToolActionHandlerData<'a>> for Fill {
+	fn process_action(&mut self, action: ToolMessage, data: ToolActionHandlerData<'a>, responses: &mut VecDeque<Message>) {
+		let mouse_pos = data.2.mouse.position;
+		let quad = [
+			glam::DVec2::new(mouse_pos.x as f64 - 1., mouse_pos.y as f64 - 1.),
+			glam::DVec2::new(mouse_pos.x as f64 + 1., mouse_pos.y as f64 - 1.),
+			glam::DVec2::new(mouse_pos.x as f64 - 1., mouse_pos.y as f64 + 1.),
+			glam::DVec2::new(mouse_pos.x as f64 + 1., mouse_pos.y as f64 + 1.),
+		];
+		if let Some(path) = data.0.intersects_quad_root(quad).get(0) {
```

Comment:
I assume you want to fill the top most layer, so `.get(0)` should be `.last()`

Reply:
Agreed, done.